Readme.md
Node Rating
===========
Rating and Hit system

Dependencies
------------

### Runtime

* Node 0.4.x+
* express 1.0.7
* connect 0.5.7
* mongoose 1.0.10  (and MongoDB server)
* cluster 0.3.2 (optional)

### Tests
* nodeunit

Running
-------
### Normal
$ node server.js 

### Clustered
$ node cluster-server.js

### Usuage

Hit insert
$ curl -X POST http://localhost:3000/hit/:context/:subject/:id

Hit count for the current hour
$ curl http://localhost:3000/hit/:context/:subject/:id

Rate insert
$ curl -X POST http://localhost:3000/rate/:context/:subject/:id

Rate count

Running tests
-------------
$ nodeunit test
or
$ node run-tests

To-Do
-----
* URIs to consume totals
* Security key
* Code coverage

Author
------

* Emerson Macedo (<http://codificando.com/>)

License:
--------

(The MIT License)

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


Changelog.md
### 0.2.7
- Bug Fix: Return 0 when GET total from not found hit search
- Bug Fix: Return 0 when GET total from not found rate search

### 0.2.6
- Using parseInt because cluster needs integer value for cluster.liste(port)
- Added production configuration to express app config

### 0.2.5
- Added cluster support

### 0.2.4
- Update database-cleaner to 0.2.0

### 0.2.3
- Write only one record for the same hour
- Update mongoose to 1.0.10

### 0.2.2
- Added rate count for thumbs up

### 0.2.1
- Extracted database cleaner to a npm package
- Using node-database-cleaner to clean mongodb after tests

### 0.2.0
- Record ratings
- Use cookie to prevent an user to rate more than one time in the same context/subject/id
- Updated mongoose to 1.0.0
- Route to get hit count

### 0.1.0
- Record hits
- Test structure 



test/hit.test.js
var testHelper = require('./helper/test-helper')
  , DatabaseCleaner = require('database-cleaner')
  , databaseCleaner = new DatabaseCleaner('mongodb')
  , testCase = require('nodeunit').testCase
  , mongoose = require('mongoose')
  , db = mongoose.createConnection('mongodb://localhost/rating_test')
  , Hit = require('hit').Hit(db);

module.exports = testCase({
  setUp: function (callback) {
    callback();
  },
  tearDown: function (callback) {
    databaseCleaner.clean(mongoose.createConnection('mongodb://localhost/rating_test').db, callback);
  },
  'It should have correct context, subject and id': function (test) {
    var hitData = {context: 'video', subject: 'media', id: '1234', date: testHelper.fakeDate()};
    Hit.hit(hitData, function(err) {
      test.strictEqual(err, null);
      test.done();
    });
  },
  'it should write only one record for the current hour': function (test) {
    var currentHour = testHelper.fakeDate();

    Hit.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Hit.hit({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err){
        test.ok(!err, "should not raise error");
        Hit.count({context: 'video', subject: 'media', id: '1234', date: currentHour}, function(err, count) {
          test.equal(count, 1, "should have only one record");
          test.done();
        });
      });
    });
  },
  'it should write different records when different hour': function (test) {
    var currentHour = testHelper.fakeDate(10);
    var nextHour = testHelper.fakeDate(11);

    Hit.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Hit.hit({context: 'video', subject: 'media', id: '1234', date: nextHour}, function (err){
        test.ok(!err, "should not raise error");
        Hit.count({context: 'video', subject: 'media', id: '1234'}, function(err, count) {
          test.equal(count, 2, "should have two record");
          test.done();
        });
      });
    });
  },
  'it should increment total when the same hour': function (test) {
    var currentHour = testHelper.fakeDate();

    Hit.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Hit.hit({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err){
        test.ok(!err, "should not raise error");
        Hit.total({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err, total) {
          test.equal(total, 2, "should have two record");
          test.done();
        });
      });
    });
  },
  'it should not increment total when the different hour': function (test) {
    var currentHour = testHelper.fakeDate(10);
    var nextHour = testHelper.fakeDate(11);

    Hit.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Hit.hit({context: 'video', subject: 'media', id: '1234', date: nextHour}, function (err){
        test.ok(!err, "should not raise error");
        Hit.total({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err, total) {
          test.equal(total, 1, "should have two record");
          test.done();
        });
      });
    });
  },
  'should return 0 when not found': function(test) {
    var currentHour = testHelper.fakeDate(10);
    Hit.total({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err, total) {
      test.equal(total, 0, "should have no record");
      test.done();
    });
  },
  'it should require date': function(test) {
    var hitData = {context: 'video', subject: 'media', id: '1234'};
    Hit.hit(hitData, function (err) {
      test.ok(err, "should raise validation error");
      test.done();
    });
  }
});


test/rate.test.js
var testHelper = require('./helper/test-helper')
  , DatabaseCleaner = require('database-cleaner')
  , databaseCleaner = new DatabaseCleaner('mongodb')
  , testCase = require('nodeunit').testCase
  , mongoose = require('mongoose')
  , db = mongoose.createConnection('mongodb://localhost/rating_test')
  , Rate = require('rate').Rate(db);

module.exports = testCase({
  setUp: function (callback) {
    callback();
  },
  tearDown: function (callback) {
    databaseCleaner.clean(mongoose.createConnection('mongodb://localhost/rating_test').db, callback);
  },
  'It should have correct context, subject and id': function (test) {
    var rateData = {context: 'video', subject: 'media', id: '1234', date: testHelper.fakeDate()};
    Rate.rate(rateData, function(err) {
      test.strictEqual(err, null);
      test.done();
    });
  },
  'it should write only one record for the current hour': function (test) {
    var currentHour = testHelper.fakeDate();

    Rate.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Rate.rate({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err){
        test.ok(!err, "should not raise error");
        Rate.count({context: 'video', subject: 'media', id: '1234', date: currentHour}, function(err, count) {
          test.equal(count, 1, "should have only one record");
          test.done();
        });
      });
    });
  },
  'it should write different records when different hour': function (test) {
    var currentHour = testHelper.fakeDate(10);
    var nextHour = testHelper.fakeDate(11);

    Rate.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Rate.rate({context: 'video', subject: 'media', id: '1234', date: nextHour}, function (err){
        test.ok(!err, "should not raise error");
        Rate.count({context: 'video', subject: 'media', id: '1234'}, function(err, count) {
          test.equal(count, 2, "should have two record");
          test.done();
        });
      });
    });
  },
  'it should increment total when the same hour': function (test) {
    var currentHour = testHelper.fakeDate();

    Rate.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Rate.rate({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err){
        test.ok(!err, "should not raise error");
        Rate.total({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err, total) {
          test.equal(total, 2, "should have two record");
          test.done();
        });
      });
    });
  },
  'it should not increment total when the different hour': function (test) {
    var currentHour = testHelper.fakeDate(10);
    var nextHour = testHelper.fakeDate(11);

    Rate.collection.insert({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err) {
      Rate.rate({context: 'video', subject: 'media', id: '1234', date: nextHour}, function (err){
        test.ok(!err, "should not raise error");
        Rate.total({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err, total) {
          test.equal(total, 1, "should have one record");
          test.done();
        });
      });
    });
  },
  'should return 0 when not found': function(test) {
    var currentHour = testHelper.fakeDate(10);
    Rate.total({context: 'video', subject: 'media', id: '1234', date: currentHour}, function (err, total) {
      test.equal(total, 0, "should have no record");
      test.done();
    });
  },
  'it should require date': function(test) {
    var rateData = {context: 'video', subject: 'media', id: '1234'};
    Rate.rate(rateData, function (err) {
      test.ok(err, "should raise validation error");
      test.done();
    });
  }
});


test/node-rating.test.js
var testHelper = require('./helper/test-helper')
  , app = require('../node-rating')
  , mongoose = require('mongoose')
  , DatabaseCleaner = require('database-cleaner')
  , databaseCleaner = new DatabaseCleaner('mongodb')
  , testCase = require('nodeunit').testCase
  , testPort = '9999';

module.exports = testCase({
  setUp: function (callback) {
    app.listen(testPort);
      this.requestParams = {
        host: 'localhost'
      , port: testPort
      , method: 'GET'
      , body: ''
      , headers: {
          contentType: 'text/plain'
      }
    };
    callback();
  },
  tearDown: function (callback) {
    app.close();
    databaseCleaner.clean(mongoose.createConnection('mongodb://localhost/rating_test').db);
    if (this.hitStub) this.hitStub.restore();
    if (this.rateStub) this.rateStub.restore();
    callback();
  },
  'should return 500 when can not hit': function (test) {
    this.hitStub = testHelper.stub(app.Hit, {
      save: function(callback) {callback.apply(this, [{}, null]);}
    });
    this.requestParams.uri = '/hit/videos/media/123';
    this.requestParams.method = 'POST';

    testHelper.makeRequest(this.requestParams, function(response) {
      test.equals (response.statusCode, 500);
      test.done();
    });
  },
  'should return 200 when hit': function (test) {
    this.requestParams.uri = '/hit/videos/media/123';
    this.requestParams.method = 'POST';
    
    testHelper.makeRequest(this.requestParams, function(response) {
      test.equals (response.statusCode, 200);
      test.done();
    });
  },
  'should return 200 when rate': function(test) {
    this.requestParams.uri = '/rate/videos/media/123';
    this.requestParams.method = 'POST';

    testHelper.makeRequest(this.requestParams, function(response) {
      test.equals (response.statusCode, 200);
      var regexp_cookie = /rate_videos_media_123=true/;
      test.ok(regexp_cookie.test(response.headers['set-cookie']));
      test.done();
    });
  },
  'should return 400 when rate more than 1 time': function(test) {
    this.requestParams.uri = '/rate/videos/media/123';
    this.requestParams.method = 'POST';
    this.requestParams.headers.cookie = 'rate_videos_media_123=rate';

    testHelper.makeRequest(this.requestParams, function(response) {
      test.equals (response.statusCode, 400);
      test.done();
    });
  },
  'should return 500 when can not rate': function (test) {
    this.rateStub = testHelper.stub(app.Rate, {
      save: function(callback) {callback.apply(this, [{}, null]);}
    });
    this.requestParams.uri = '/rate/videos/media/123';
    this.requestParams.method = 'POST';

    testHelper.makeRequest(this.requestParams, function(response) {
      test.equals (response.statusCode, 500);
      test.done();
    });
  },
  'should get hit count given an element': function(test) {
    var requestParams = this.requestParams;
    requestParams.uri = '/hit/videos/media/123';
    requestParams.method = 'POST';

    testHelper.makeRequest(requestParams, function(response) {
      testHelper.makeRequest(requestParams, function(response) {
        requestParams.method = 'GET';
        testHelper.makeRequest(requestParams, function(response) {
          test.equals (response.body, '2');
          test.done();
        });
      });
    });
  },
  'should get thumbs up count given an element': function(test) {
    var requestParams = this.requestParams;
    requestParams.uri = '/rate/videos/media/123';
    requestParams.method = 'POST';

    testHelper.makeRequest(requestParams, function(response) {
      testHelper.makeRequest(requestParams, function(response) {
        requestParams.method = 'GET';
        testHelper.makeRequest(requestParams, function(response) {
          test.equals (response.body, '2');
          test.done();
        });
      });
    });
  },
  'should return 0 when getting rate count for an inexistent element': function(test) {
    var requestParams = this.requestParams;
    requestParams.uri = '/rate/videos/media/123';
    requestParams.method = 'GET';
    testHelper.makeRequest(requestParams, function(response) {
      test.equals (response.body, '0');
      test.done();
    });
  },
  'should return 0 when getting hit count for an inexistent element': function(test) {
    var requestParams = this.requestParams;
    requestParams.uri = '/hit/videos/media/123';
    requestParams.method = 'GET';
    testHelper.makeRequest(requestParams, function(response) {
      test.equals (response.body, '0');
      test.done();
    });
  }
});


lib/hit.js
var mongoose = require('mongoose')
  , Hit;

Hit = new mongoose.Schema({
    context: { type: String, index: true}
  , subject: { type: String, index: true}
  , id: { type: String, index: true}
  , date: {type: Date, index:true}
  , total: {type: Number, default: 1}
});
var util = require('util');
Hit.static('hit', function(data, callback) {
    if (!data.date) {
      callback(true);
      return;
    }
    var HitModel = this;
    this.findOne(data, function (err, hit) {
      if (hit) {
        hit.total++;
        hit.save(callback);
      } else {
        hit = new HitModel(data);
        hit.save(callback);
      }
    });
});

Hit.static('total', function (data, callback) {
  this.findOne(data, function (err, hit) { 
      if (hit) {
        callback(null, hit.total.valueOf());   
      } else {
        callback(null, 0); 
      }
  });
});

mongoose.model('Hit', Hit);
   
exports.Hit = function(db) {
  return db.model('Hit');
}


lib/rate.js
var mongoose = require('mongoose')
  , Rate;

Rate = new mongoose.Schema({
    context: { type: String, index: true}
  , subject: { type: String, index: true}
  , id: { type: String, index: true}
  , date: {type: Date, index:true}
  , total: {type: Number, default: 1}
});
var util = require('util');
Rate.static('rate', function(data, callback) {
    if (!data.date) {
      callback(true);
      return;
    }
    var RateModel = this;
    this.findOne(data, function (err, rate) {
      if (rate) {
        rate.total++;
        rate.save(callback);
      } else {
        rate = new RateModel(data);
        rate.save(callback);
      }
    });
});

Rate.static('total', function (data, callback) {
  this.findOne(data, function (err, rate) { 
      if (rate) {
        callback(null, rate.total.valueOf());   
      } else {
        callback(null, 0);
      }
  });
});

mongoose.model('Rate', Rate);
   
exports.Rate = function(db) {
  return db.model('Rate');
}


node-rating.js
require.paths.unshift('./lib');
var express = require('express')
  , app = express.createServer()
  , db
  , mongoose = require('mongoose');

var currentHour = function() {
  var hour = new Date();
  hour.setMinutes(0);
  hour.setSeconds(0);
  hour.setMilliseconds(0);
  return hour;
};

app.configure(function () {
  app.use(express.cookieDecoder());
  app.use(express.bodyDecoder());
  app.use(express.staticProvider(__dirname + '/public'));
});

app.configure('development', function() {
  app.use(express.logger());
  app.use(express.errorHandler({ dumpExceptions: true, showStack: true }));
  app.set('db-uri', 'mongodb://localhost/rating_development');
});

app.configure('test', function() {
  app.use(express.errorHandler({ dumpExceptions: true, showStack: true }));
  app.set('db-uri', 'mongodb://localhost/rating_test');
});

app.configure('production', function() {
  app.use(express.logger());
  app.use(express.errorHandler());
  app.set('db-uri', 'mongodb://localhost/rating');
});

db = mongoose.createConnection(app.set('db-uri'));
app.Hit = require('hit').Hit(db);
app.Rate = require('rate').Rate(db);

app.get('/hit/:context/:subject/:id', function(req, res) {
  var queryData = {
      date: currentHour()
    , context: req.params.context
    , subject: req.params.subject
    , id: req.params.id };

  app.Hit.total(queryData, function (err, count) {
     res.send(count.toString(), 200);
  });

});

app.get('/rate/:context/:subject/:id', function(req, res) {
  var queryData = {
      context: req.params.context
    , subject: req.params.subject
    , id: req.params.id };

  app.Rate.total(queryData, function (err, count) {
     res.send(count.toString(), 200);
  });
});

app.post('/hit/:context/:subject/:id', function(req, res) {
  var hitData = {context: req.params.context, subject: req.params.subject, id: req.params.id, date: currentHour()};
  app.Hit.hit(hitData, function (err) {
    if (err) { res.send(500); res.end(); }
    res.send(200);
  });
});

app.post('/rate/:context/:subject/:id', function(req, res) {
  var cookieName = "rate_" + req.params.context + "_" + req.params.subject + "_" + req.params.id;

  if (cookieName in req.cookies) {
    res.send(400);
    res.end();
  }

  var rateData = {context: req.params.context, subject: req.params.subject, id: req.params.id, date: currentHour()};
  app.Rate.rate(rateData, function (err) {
    if (err) { res.send(500); res.end(); }
    res.cookie(cookieName, "true", { expires: new Date(Date.now() + (1000 * 60 * 60 * 24 * 365)), httpOnly: true });
    res.send(200);
  });
});
module.exports = app;
